### ISSUE TITLE
Memory leak during repeated runs on node.js?

### ISSUE THREAD

**Issue by apowers313** [emscripten-core/emscripten#12740]
Running on node v14.8.0, when I run the same emscripten function repeatedly I eventually run out of memory. Here's a gist that shows the memory ticking up by 16MB each run for a minimal C file that doesn't allocate any memory:
https://gist.github.com/apowers313/cee96b41e1f3ba272b214fc7e1c983bb

**Comment by kripken** [comment]
One possible issue is that `quit()` is expected to actually quit the problem (it normally throws an exception or calls exit, I believe).

If it's not that, it may be a leak in node. Running in different browsers can be a good way to see if the problem is just in node and/or just in v8. If the problem does show up in a browser, the devtools can help find the leak.

**Comment by apowers313** [comment]
If I recompile with `-s NO_EXIT_RUNTIME` and use the `postRun` hook rather than `quit`, I get the exact same memory leak. 

Note that the leak shows up in the `external` property from [process.memoryUsage()](https://nodejs.org/api/process.html#process_process_memoryusage) which "refers to the memory usage of C++ objects bound to JavaScript objects managed by V8". Makes me think that you're right -- it might be a node / V8 / Chromium problem.

Maybe [this Chromium bug](https://bugs.chromium.org/p/v8/issues/detail?id=11034&q=WebAssembly%20leak&can=2) is related? Are all y'all using the WASM C++ API?

**Comment by kripken** [comment]
No. Emscripten just emits WebAssembly and JavaScript, it doesn't use any special APIs at any other level. Just things that also work on the Web.

**Comment by apowers313** [comment]
In retrospect, that was kindof a dumb question. :)

I found the leak though. If I add `process.removeAllListeners("uncaughtException")` and `process.removeAllListeners("unhandledRejection")` to my code then it will run for hundreds of thousands of loops without leaking memory.

Maybe emscripten should clean up those handlers when it's done running in the node environment?

**Comment by curiousdannii** [comment]
I was guessing that it could've been a lingering event handler... 16MB sounds like the wasm memory allocation, but then it's kind of odd that it's *only* leaking 16MB, because there should be a lot of other stuff in the closure using memory.

It definitely should clean up the handlers.

**Comment by kripken** [comment]
Good points, yes, we should fix this.

Does changing the `on()` calls to `once()` fix it? Seems like it should. If so I can open a PR.

**Comment by kripken** [comment]
Oh, no, that would only help if they are actually called. Then I'm not sure offhand where these would be unregistered. Would be good if someone has time to look into it, adding some labels to the issue.

**Comment by apowers313** [comment]
Two thoughts for options:
1. Add a `cleanUp()` function to `callMain()`; or
2. Call `addOnPostRun()` to add cleanup for node.js?

**Comment by curiousdannii** [comment]
Maybe this just needs documenting, that if you intend to call the factory multiple times turn off `NODEJS_CATCH_EXIT` and `NODEJS_CATCH_REJECTION`? Or could they be made off-by-default? Node 15 has been published so the second is only needed for older Node versions (though Node 15 isn't LTS so you may want to wait.)

**Comment by curiousdannii** [comment]
Actually I think it would be better to clean up if possible. The `uncaughtException` handler isn't an issue as it doesn't capture anything. It's just `unhandledRejection` which captures the `abort` function.

It's tricky because we can't just unregister the event hook in an exit function or the like. A user may be running a module in non-exiting mode, letting it go out of scope, which would normally let the whole module be garbage collected, but with the event handler it has to be kept. Though looking at the code for `abort` I'm not actually sure how it all sticks around...

One possibility: Use a `WeakMap` to decouple the event handler from `abort` so that the module can be GCed without needing to unregister the hook. (`WeakRef` would be neater but it's not supported enough.) It would still leak memory, but so much less.

**Comment by kripken** [comment]
Good point that it's just for `unhandledRejection`. In that case this is less urgent to fix as once node fixes the bug in an LTS release I think we can remove that workaround.

But if the WeakMap idea works it sounds good to me!

**Comment by curiousdannii** [comment]
So turns out you already had that idea: #9820

**Comment by kripken** [comment]
Heh, I'd forgotten about it too, until the bot reminded... Yeah, it would be good to do it, if someone has time.

**Comment by sbc100** [comment]
bump
